🚀 1단계 - 스프링 프레임워크 적용

Chiptune93 opened the pull request:
step1 최초 기본 작업 후, 리뷰 요청 드립니다

Commit:
feat($all): first commit, coding all functional class and others

controller, service, dao, model, static pages, database modified and added
read README.md

hyunssooo commented on src/main/java/racingcar/Controller/RacingCarController.java:
```diff
@@ -0,0 +1,32 @@
+
+@RestController
+public class RacingCarController {
+
+    @Autowired
+    RacingService svc;
+
```
생성자로 주입 받는 것은 어떨까요? 
`@Autowired` 를 생략할 수 있습니다. 그리고 접근제어자가 빠졌네요. `svc` 라는 이름보다 명확한 이름을 지어주는 것은 어떨까요?

hyunssooo commented on src/main/java/racingcar/Controller/RacingCarController.java:
```diff
@@ -0,0 +1,32 @@
+    RacingService svc;
+
+    @PostMapping("/plays")
+    public ResponseEntity<PlayResultOut> racingStart(@RequestBody PlayResultIn pri) {
+        return new ResponseEntity<>(svc.racing(pri.getNames(), pri.getCount()),
+            HttpStatus.OK);
+    }
```
pir 라는 이름은 다른 사람이 읽기에 도움이 될까요?

hyunssooo commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,75 @@
+package racingcar.model;
+
+public class RacingCar {
+
```
model 패키지 아래에 있는 클래스는 전부 setter가 존재하는데 이 부분을 제거해 보시면 좋을 것 같습니다.

Chiptune93 replied:
사실 getter & setter 를 자동 생성을 통해 생성하다보면 setter는 자연스럽게 들어가는데요.
로직 중, 생성자를 통해 만들어진 모델에 setter를 통해 특정 값만 세팅 하고자 할 때
setter가 없으면 어떤 방법을 통해 특정 값을 해당 모델에 추가하는 방법을 사용할까요?

참고할만한 내용이나 글 있으면 추천 부탁드릴게요!

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,94 @@
+        Random random = new Random();
+        for (RacingCar rc : racingCarList) {
+            int randomNumber = random.nextInt(10);
+            System.out.println(rc.getName() + ":" + rc.getGroupId() + "/" + randomNumber);
+            racingCarDAO.updatePosition(rc, randomNumber);
```
데이터를 보시려면 출력보다 로그를 사용해 보면 좋을 것 같아요.

Chiptune93 replied:
피드백 주신 내용 참고하여 log4j 모듈 추가해 커밋하겠습니다

hyunssooo commented on src/main/java/racingcar/RacingCar.java:
```diff
@@ -0,0 +1,25 @@
+package racingcar;
+
+class RacingCar {
```
이 클래스와 model 패키지 하위의 RacingCar는 어떻게 다른가요? 이 클래스를 그대로 사용해도 괜찮을 것 같아서요.

Chiptune93 replied:
기존 레거시 코드들을 가져오면서 남긴 파일이고, 사실 모델에 있는 것도 이것을 그대로 가져가 사용했습니다.
지워야 한다면 지우겠지만 우선 의미는 그렇습니다.
다음 피드백 때 지우는 것을 추천해주시면 지우도록 하겠습니다~

hyunssooo reviewed (changes requested):
안녕하세요. 
이번 미션 함께 진행할 정현수입니다. 
많이 늦어서 죄송해요. 🥲
몇 가지 코멘트 남겼습니다. 확인 부탁드려요!

Commit:
maintain($all): feedback applied

added feature
- log4j : for logging database and custom debugging

maintain files : feedback applied
- dao
- service
- model

Chiptune93 commented:
안녕하세요~ 현수님!

피드백 주신 내용을 바탕으로 수정한 코드를 다시 커밋 하였습니다.
확인 부탁 드립니다! 또한, 기존 피드백 내용에도 댓글을 몇 개 달았는데
그것도 같이 확인 부탁 드릴게요!

좋은 피드백 감사드립니다!

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+
+@Service
+public class RacingService {
+
+    @Autowired
+    PersonDAO personDAO;
+
+    @Autowired
+    PlayResultDAO playResultDAO;
+
+    @Autowired
+    RacingCarDAO racingCarDAO;
```
https://github.com/next-step/spring-racingcar/pull/3#discussion_r1147113256

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+    public List<PlayResultOut> playList() {
+        List<PlayResultOut> result = new ArrayList<>();
+        List<PlayResult> list = playResultDAO.findAllPlayResult();
+        for (PlayResult pr : list) {
+            List<RacingCar> racingCarByGroupId = racingCarDAO.findRacingCarByGroupId(
+                pr.getGroupId());
+            PlayResultOut pro = new PlayResultOut(pr.getWinners(), racingCarByGroupId);
+            result.add(pro);
+        }
```
코드 전체적으로 축약어를 많이 사용하고 있는데요. 
https://edu.nextstep.camp/s/jmQRkjTB/ls/rvaYDE9l
축약어는 코드를 읽는 사람에게 도움이 되지 않을 것 같아요.

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+    public PlayResultOut racing(String inputName, int inputTryNumber) {
+        int groupId = racingCarDAO.getGroupId();
+        String[] names = inputName.split(",");
+        for (String name : names) {
+            Person person = new Person(name);
```
Person 은 꼭 필요한가요? 어떤 역할과 책임을 가지고 있나요?

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @param groupId
+     */
+    public void playRound(int groupId) {
+        List<RacingCar> racingCarList = racingCarDAO.findRacingCarByGroupId(groupId);
+        Random random = new Random();
+        for (RacingCar rc : racingCarList) {
+            int randomNumber = random.nextInt(10);
+            logger.debug(rc.getName() + ":" + rc.getGroupId() + "/" + randomNumber);
+            racingCarDAO.updatePosition(rc, randomNumber);
+        }
+    }
```
RacingCar 의 위치를 변경하는 것이 아니라 DB에 바로 업데이트를 하고 있네요. 
저번에 진행했던 자동차 경주 미션을 떠올려 객체지향 프로그래밍을 진행해 보는 것이 어떨까요??

hyunssooo commented on src/main/java/racingcar/model/PlayResultOut.java:
```diff
@@ -0,0 +1,23 @@
+package racingcar.model;
+
+import java.util.List;
+
+public class PlayResultOut {
+
```
dto로 보이네요 dto도 패키지를 따로 분리해 보는 것은 어떨까요? 
어떤 기준으로 model이라는 패키지를 만드셨나요? 생각하시는 model은 무엇인가요??

hyunssooo commented on src/main/java/racingcar/model/RacingCar.java:
```diff
@@ -0,0 +1,52 @@
+    private int personId;
+    private String name;
+    private int position;
+    private String createdAt;
+
```
사용하고 있지 않은 필드 같은데 제거하는 것은 어떨까요??

hyunssooo commented on src/main/java/racingcar/Repository/RacingCarDAO.java:
```diff
@@ -0,0 +1,131 @@
+     * @param racingCar
+     * @param randomNumber
+     */
+    public void updatePosition(RacingCar racingCar, int randomNumber) {
+        if (randomNumber >= 4) {
+            String sql = "update racing_car set position = ? where group_id = ? and person_id = ?";
+            this.jdbcTemplate.update(sql, racingCar.getPosition() + 1, racingCar.getGroupId(),
+                racingCar.getPersonId());
+        }
+    }
```
https://github.com/next-step/spring-racingcar/pull/3#discussion_r1148364846 이 질문과 이어질 것 같아요. 

`숫자가 4 이상이면 자동차 위치가 변경된다` 이 부분은 핵심 비즈니스라고 볼 수 있습니다. 이 경우는 도메인 모델 내부에 비즈니스 로직이 있는 것이 맞을 것 같아요. 지난번에 진행한 자동차 경주 미션과 로또 미션 처럼 비즈니스를 담당하는 도메인 모델 내부에 비즈니스 로직을 작성하는 것은 어떨까요??

hyunssooo reviewed (changes requested):
안녕하세요. 
몇 가지 코멘트 남겼습니다! 
궁금하시거나 함께 고민하고 싶은 부분이 있으시면 dm이나 코멘트 남겨주세요!

Commit:
maintain($all): feedback applied #2

maintain files : feedback applied
- service
- dao
- position update DB -> Object

new package
- dto

Commit:
maintain($all): feedback applied #3

remove unused test file

Chiptune93 commented:
안녕하세요~ 현수님!

피드백 주신 내용을 바탕으로 수정한 코드를 다시 커밋 하였습니다.

### https://github.com/next-step/spring-racingcar/pull/3#discussion_r1148495692
- Person 객체의 경우 기본 제공된 프로젝트에 생성 예시에 있어 같이 구현하였습니다. 다만 기존에 진행했던 로직에서는 필요없던 객체가 맞아 제외하였습니다.

### https://github.com/next-step/spring-racingcar/pull/3#discussion_r1148496222
- dto 객체가 맞으나 저 같은 경우 모델과 항상 같이 놔두던 습관 때문에 구성이 그렇게 되었습니다. 말씀해주신대로 패키지를 따로 분리하였습니다.

### https://github.com/next-step/spring-racingcar/pull/3#discussion_r1148496611
- 기존에 진행했던 자동차 경주 자바 버전 처럼, 객체에서 처리하고 경주 후 마지막에 전체 업데이트 하는 것으로 변경하였습니다. 

좋은 피드백 감사드립니다!

hyunssooo commented on README.md:
```diff
@@ -1 +1,108 @@
 # jwp-racingcar
+
+# Step1 : Spring Framework applied
+
```
readme 파일도 최신화가 필요해 보이네요 :)

hyunssooo commented on src/main/java/racingcar/RacingCar.java:
```diff
@@ -0,0 +1,25 @@
+        return this.position;
+    }
+
+    public void move(int randomNumber) {
+        if (randomNumber >= 4) {
+            this.position++;
+        }
+    }
```
객체 스스로 판단하게 한 점 💯 
더 이상 Car 객체는 랜덤한 숫자에 의존하지 않는데요. 그럼 Random 이라는 단어는 빠져도 괜찮지 않을까요? 
간단한 클래스이지만 테스트도 함께 작성되면 더욱 좋을 것 같아요.

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+@Service
+public class RacingService {
+
+    final
+    PlayResultDAO playResultDAO;
+
+    final
+    RacingCarDAO racingCarDAO;
```
```suggestion
    private final PlayResultDAO playResultDAO;
    private final RacingCarDAO racingCarDAO;
```

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+     * @param racingCarList
+     * @param groupId
+     */
+    public void playRound(List<RacingCar> racingCarList, int groupId) {
+        Random random = new Random();
+        for (RacingCar racingCar : racingCarList) {
+            int randomNumber = random.nextInt(10);
+            logger.debug(racingCar.getName() + ":" + racingCar.getGroupId() + "/" + randomNumber);
+            racingCar.move(randomNumber);
+            //racingCarDAO.updatePosition(rc, randomNumber);
+        }
+    }
```
groupId 는 사용하지 않네요! 
사용하지 않는 코드는 주석보다 제거하는 것이 어떨까요?

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+        }
+        racingCarDAO.updateRacingCarList(racingCarList);
+
+        List<RacingCar> winnerList = racingCarDAO.getWinner(groupId);
+        List<String> winners = new ArrayList<>();
```
List를 사용하는데 변수명이 List로 끝나는 경우도 있고 복수형으로 끝나는 경우도 있네요. 
이 부분도 통일해 보면 어떨까요? 
개인적으로 `List<RacingCar>` 의 관심사는 List 보다 RacingCar 여러개라고 생각해요. 그래서 저는 보통 복수형으로 작성하는 편입니다. 
또한 List에서 Set으로 변경된다면 참조하고 있는 부분은 모두 수정이 필요하겠네요.

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+            playRound(racingCarList, groupId);
+        }
+        racingCarDAO.updateRacingCarList(racingCarList);
+
+        List<RacingCar> winnerList = racingCarDAO.getWinner(groupId);
```
우승자를 구하는 것은 핵심 비즈니스 아닐까요? 이 부분이 DB에 의존하고 있어요. 도메인 모델에서 처리하는 방법은 없을까요?

hyunssooo replied:
`RacingCars` 와 같은 일급컬렉션을 만들어 보는 것은 어떨까요? 
https://tecoble.techcourse.co.kr/post/2020-05-08-First-Class-Collection/
이 글이 도움이 되면 좋겠어요.

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+        for (RacingCar rc : winnerList) {
+            winners.add(rc.getName());
+        }
+        playResult.setWinners(String.join(", ", winners));
+        playResultDAO.updatePlayResult(playResult);
```
playResult는 생성과 업데이트가 각각 한번씩 이루어 지는데 이 부분을 한번으로 줄여볼 수 없을까요? 
RacingCar들이 모두 move하고 playResult를 생성하는 방법은 어떨까요?

hyunssooo commented on src/main/java/racingcar/Service/RacingService.java:
```diff
@@ -0,0 +1,101 @@
+        }
+        racingCarDAO.updateRacingCarList(racingCarList);
+
+        List<RacingCar> winnerList = racingCarDAO.getWinner(groupId);
+        List<String> winners = new ArrayList<>();
+        for (RacingCar rc : winnerList) {
+            winners.add(rc.getName());
+        }
```
여기는 stream을 활용해 봐도 좋을 것 같아요. 
```java
        List<RacingCar> winners = racingCarDAO.getWinner(groupId);
        final List<String> winnerNames = winners.stream()
            .map(RacingCar::getName)
            .collect(Collectors.toList());
```

hyunssooo reviewed (approved):
안녕하세요. 
몇 가지 코멘트 남겼습니다! 
다음 step 진행하시면서 확인 부탁드릴게요! 
다음 step에서 뵙겠습니다!

Chiptune93 commented:
안녕하세요, 현수님!

마지막까지 피드백 감사드립니다.
보내주신 내용은 STEP2 브랜치 생성 후, 피드백 적용하였습니다.

다음 미션 때 뵐게요!
